Physics/Dynamics/Voigtian.hpp

```diff
@@ -42,106 +35,74 @@ namespace Dynamics {
-
-  /// Get meson radius.
-  /// The meson radius is a measure of the size of the resonant state. It is
-  /// used to calculate the angular momentum barrier factors.
-  double GetMesonRadius() const { return MesonRadius->value(); }
-
-  /// \see GetFormFactorType()
-  void SetFormFactorType(FormFactorType t) { FFType = t; }
-
-  /// Get form factor type.
-  /// The type of formfactor that is used to calculate the angular momentum
-  /// barrier factors.
-  FormFactorType GetFormFactorType() { return FFType; }
-
-  void SetSigma(double sigma) { Sigma = sigma; }
-
-  double GetSigma() const { return Sigma; }
-
-  void updateParametersFrom(const ComPWA::FunctionTree::ParameterList &list);
-  void addUniqueParametersTo(ComPWA::FunctionTree::ParameterList &list);
-  void addFitParametersTo(std::vector<double> &FitParameters) final;
-
-  std::shared_ptr<ComPWA::FunctionTree::FunctionTree>
-  createFunctionTree(const ComPWA::FunctionTree::ParameterList &DataSample,
-                     unsigned int pos, const std::string &suffix) const;
-
-protected:
-  /// Orbital Angular Momentum between two daughters in Resonance decay
-  ComPWA::Spin L;
-
-  /// Masses of daughter particles
-  std::pair<double, double> DaughterMasses;
-
-  /// Names of daughter particles
-  std::pair<std::string, std::string> DaughterNames;
-
-  /// Resonance mass
-  std::shared_ptr<ComPWA::FunctionTree::FitParameter> Mass;
-
-  /// Decay width of resonante state
-  std::shared_ptr<ComPWA::FunctionTree::FitParameter> Width;
-
-  /// Meson radius of resonant state
-  std::shared_ptr<ComPWA::FunctionTree::FitParameter> MesonRadius;
-
-  /// Form factor type
-  FormFactorType FFType;
+struct InputInfo : RelativisticBreitWigner::InputInfo {
   /// resolution: the width of gaussian function which is used to represent the
   /// resolution of mass spectrum
   double Sigma;
 };
 
+/// Dynamical voigt function.
+/// \param mSq Invariant mass squared
+/// \param mR Mass of the resonant state
+/// \param wR Width of the resonant state
+/// \param sigma Width of the gaussian, i.e., the resolution of the mass
+/// spectrum at mR \return Amplitude value
+inline std::complex<double> dynamicalFunction(double mSq, double mR, double wR,
+                                              double sigma) {
+
+  double sqrtS = sqrt(mSq);
+
+  // the non-relativistic BreitWigner which is convoluted in Voigtian
+  // has the exactly following expression:
+  // BW(x, m, width) = 1/pi * width/2 * 1/((x - m)^2 + (width/2)^2)
+  // i.e., the Lorentz formula with Gamma = width/2 and x' = x - m
+  /// https://root.cern.ch/doc/master/RooVoigtianian_8cxx_source.html
+  double argu = sqrtS - mR;
+  double c = 1.0 / (sqrt(2.0) * sigma);
+  double a = c * 0.5 * wR;
+  double u = c * argu;
+  std::complex<double> z(u, a);
+  std::complex<double> v = Faddeeva::w(z, 1e-13);
+  double val = c * 1.0 / sqrt(M_PI) * v.real();
+  double sqrtVal = sqrt(val);
+
+  /// keep the phi angle of the complex BW
```

Comment:
Now I think this thought (keep the phi angle of the complex BW) is too naive and meaningless, and it may also cause difficulty to understand what the voigtian is done. So I think it is better to just return a real number for this voigtian function: std::complex<double> result(sqrtVal, 0)

Reply:
Hmm, why can you remove the phase?

Reply:
We can't. But the true phase from the Breit-Wigner is not available due to the resolution. And it is not clear how to convolute a complex function with a resolution function. In the old code, I assume that the smeared function (e.g., |BW|^2oxGaus) as a complex function has the same phase with the original BW. Usually this is not true (even as an approximation it is not true in some cases, e.g., pi0/eta/D has a width at keV level, resolution at MeV level). So I think remove this assumption is better.

Reply:
Puh, ok. Should we make a separate issue + pull request for that?

Reply:
Yes, another issue and pull request will be more clear.

Reply:
Ok, I leave that to you :)